src/math/sieve_of_eratosthenes.rs

```diff
@@ -1,53 +1,108 @@
-    let mut sieve: Vec<bool> = vec![true; num + 1];
+    result
+}
 
-    while start <= end {
+/// Marks non-prime numbers in the sieve and collects prime numbers up to `end`.
+///
+/// # Arguments
+///
+/// * `sieve` - A mutable slice of booleans representing the sieve.
+/// * `end` - The square root of the upper limit, used to optimize the algorithm.
+/// * `num` - The upper limit up to which to mark non-prime numbers.
+/// * `result` - A mutable vector to store the prime numbers.
+fn update_sieve(sieve: &mut [bool], end: usize, num: usize, result: &mut Vec<usize>) {
+    for start in 2..=end {
         if sieve[start] {
-            result.push(start);
-            for i in (start * start..num + 1).step_by(start) {
-                if sieve[i] {
-                    sieve[i] = false;
-                }
+            result.push(start); // Collect prime numbers up to `end`
+            for i in (start * start..=num).step_by(start) {
+                sieve[i] = false;
             }
         }
-        start += 1;
-    }
-    for (i, item) in sieve.iter().enumerate().take(num + 1).skip(end + 1) {
-        if *item {
-            result.push(i)
-        }
     }
-    result
+}
+
+/// Extracts remaining prime numbers from the sieve beyond the given start index.
+///
+/// # Arguments
+///
+/// * `sieve` - A slice of booleans representing the sieve with non-prime numbers marked as false.
+/// * `start` - The index to start checking for primes (inclusive).
+///
+/// # Returns
+///
+/// A vector containing all remaining prime numbers extracted from the sieve.
+fn extract_remaining_primes(sieve: &[bool], start: usize) -> Vec<usize> {
+    sieve[start..]
+        .iter()
+        .enumerate()
+        .filter_map(|(i, &is_prime)| if is_prime { Some(start + i) } else { None })
+        .collect()
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
 
-    #[test]
-    fn basic() {
-        assert_eq!(sieve_of_eratosthenes(0), vec![]);
-        assert_eq!(sieve_of_eratosthenes(11), vec![2, 3, 5, 7, 11]);
-        assert_eq!(
-            sieve_of_eratosthenes(25),
-            vec![2, 3, 5, 7, 11, 13, 17, 19, 23]
-        );
-        assert_eq!(
-            sieve_of_eratosthenes(33),
-            vec![2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31]
-        );
-        assert_eq!(
-            sieve_of_eratosthenes(100),
-            vec![
-                2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79,
-                83, 89, 97
-            ]
-        );
+    macro_rules! sieve_tests {
+        ($($name:ident: $test_case:expr,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (input, expected) = $test_case;
+                    assert_eq!(sieve_of_eratosthenes(input), expected);
+                }
+            )*
+        }
+    }
+
+    sieve_tests! {
+        test_0: (0, Vec::<usize>::new()),
+        test_2: (2, vec![2]),
```

Comment:
```suggestion
        test_2: (2, vec![2]),
        test_3: (3, vec![2, 3]),
        test_4: (3, vec![2, 3]),
        // test_5: (3, vec![2, 3, 5]),
        // test_6: (3, vec![2, 3, 5]),
```
`test_5` and `test_6` fail.